index.js

```diff
@@ -4,20 +4,22 @@ const create = prod => (start, sink) => {
     sink(0, () => {});
     return;
   }
-  let started = disposed = got1 = got2 = false;
-  sink(0, t => {
-    got1 = got1 || t === 1;
-    got2 = got2 || t === 2;
-    !started && got1 && prod((pt, pd) => {
-      if (pt === 0) return;
-      started = true;
-      !disposed && got1 && pt === 1 && sink(pt, pd);
-      if ((got2 || pt === 2) && !disposed) {
-        disposed = true;
-        sink(2);
-      }
-    });
+  let started = disposed = false;
+  let unsub
+  sink(0, (t, d) => {
+    if (!started && t === 1) {
+      unsub = prod((pt, pd) => {
+        if (pt === 0) return;
+        started = true;
```

Comment:
im also a little bit unsure why producer is started upon data request (receiving 1), shouldnt it start right away (on handshake)?

Reply:
Second thought - I actually don't see why `got2` is needed 😅 

I see the same behaviour for your example with both versions of the code - I mean what's in the gist (with `got2`) and the one from this PR (without `got2`).

Immediate `unsub()` call doesn't change anything in this scenario, because data request (1) is sent before it has a chance to happen - so the producer starts and gets immediately cancelled in both cases.

I guess this still could be a problem for some other sinks - but `callbag-subscribe` imho doesnt illustrate this.

Or have I missed something?

Reply:
The producer is started right after handshake, namely inside sink(0, function); that's why we have an init variable there otherwise we'll end up with an infinite loop. If the producer is called outside the handshake `sink(0, function)` it won't work as expected.
The immediate `unsub()` clears the interval right away and the value is not emitted although the first upward value is 1.
`got2` is necessary because the producer can send downwards 2 or the sink (like `subscribe()`) can send 2 upwards and if any of those happens we want to stop emitting. You don't want the user sending sink(2) and then after that sink(1, 'a'); the 'a' should never be received by the sink after (2) is sent.
Does that make sense to you? :- )

Reply:
> The immediate unsub() clears the interval right away and the value is not emitted although the first upward value is 1.

This is correct behaviour from what I understand, mainly because `setInterval` in the example won't have a chance to emit anything before `unsub` gets called. If we put immediate `sink(1, val++)` in the producer then with both your code and mine a single value gets logged before immediate `unsub` kills the producer.

> got2 is necessary because the producer can send downwards 2 or the sink (like subscribe()) can send 2 upwards and if any of those happens we want to stop emitting. You don't want the user sending sink(2) and then after that sink(1, 'a'); the 'a' should never be received by the sink after (2) is sent.

Isnt this already handled by [this](https://github.com/Andarist/callbag-create/blob/fe908bdb11583cd712cead9de2928c332f5bf5f5/index.js#L14)?
```js
!disposed && pt === 1 && sink(pt, pd);
```

Btw. i'm still a little bit curious why producer is started [on a data request](https://github.com/Andarist/callbag-create/blob/fe908bdb11583cd712cead9de2928c332f5bf5f5/index.js#L10) and not immediately on the handshake. The first data request often comes hand in hand with a handshake, but its not always the case.